File: backend/app/services/container_ref_normalization.py
"""Container reference normalization for canonical asset identity."""

from __future__ import annotations

from dataclasses import dataclass

from app.parsers.image_digest import normalize_image_digest


def _parse_container_asset_path_aliases(raw: str) -> list[tuple[str, str]]:
    """Parse ``VAT_CONTAINER_ASSET_PATH_ALIASES``: ``from=>to;from2=>to2``."""
    out: list[tuple[str, str]] = []
    if not raw or not str(raw).strip():
        return out
    for segment in str(raw).split(";"):
        segment = segment.strip()
        if "=>" not in segment:
            continue
        left, right = segment.split("=>", 1)
        src = left.strip().lower()
        dst = right.strip().lower()
        if src and dst:
            out.append((src, dst))
    return out


def apply_container_asset_path_aliases(canonical_key: str) -> str:
    """
    Rewrite canonical registry path prefixes (tenant policy).

    Applied after ``normalize_container_ref`` so scanner ``.../operators/images/...``
    and Aikido ``.../containers/images/...`` can map to one asset key when configured.
    """
    if not canonical_key or not canonical_key.strip():
        return canonical_key
    from app.core.config import get_settings

    pairs = _parse_container_asset_path_aliases(
        get_settings().container_asset_path_aliases or ""
    )
    if not pairs:
        return canonical_key
    k = canonical_key.strip()
    lower = k.lower()
    for src, dst in pairs:
        if lower.startswith(src):
            return dst + k[len(src) :]
    return canonical_key


@dataclass(frozen=True)
class NormalizedContainerRef:
    canonical_asset_key: str
    observed_tag: str | None
    observed_digest: str | None
    raw_ref: str


def _split_digest(ref: str) -> tuple[str, str | None]:
    if "@sha256:" not in ref:
        return ref, None
    left, right = ref.split("@sha256:", 1)
    digest = normalize_image_digest(f"sha256:{right}")
    return left.strip(), digest


def _split_tag(repo_ref: str) -> tuple[str, str | None]:
    if ":" not in repo_ref:
        return repo_ref, None
    last_slash = repo_ref.rfind("/")
    last_colon = repo_ref.rfind(":")
    if last_colon > last_slash:
        tag = repo_ref[last_colon + 1 :].strip()
        if tag:
            return repo_ref[:last_colon].strip(), tag
    return repo_ref, None


def _split_registry_and_path(repo: str) -> tuple[str | None, str]:
    if not repo:
        return None, ""
    if "/" not in repo:
        return None, repo
    first, rest = repo.split("/", 1)
    if "." in first or ":" in first or first == "localhost":
        return first, rest
    return None, repo


def _docker_hub_path(path: str) -> str:
    p = (path or "").strip("/")
    if not p:
        return "library/unknown"
    if "/" not in p:
        return f"library/{p}"
    return p


def normalize_container_ref(value: str | None) -> NormalizedContainerRef:
    raw = (value or "").strip()
    if not raw:
        return NormalizedContainerRef(
            canonical_asset_key="docker.io/library/unknown",
            observed_tag=None,
            observed_digest=None,
            raw_ref="",
        )

    repo_tag_part, digest = _split_digest(raw)
    repo_part, tag = _split_tag(repo_tag_part)
    registry, path = _split_registry_and_path(repo_part)

    if registry is None:
        registry = "docker.io"
        path = _docker_hub_path(path)

    canonical = f"{registry.lower()}/{(path or '').strip('/').lower()}".strip("/")
    if not canonical:
        canonical = "docker.io/library/unknown"

    return NormalizedContainerRef(
        canonical_asset_key=canonical,
        observed_tag=(tag or None),
        observed_digest=digest,
        raw_ref=raw,
    )


def is_safe_tag_only_alias_variant(raw_ref: str | None, canonical_asset_key: str) -> bool:
    """
    True when raw_ref differs from canonical only by an explicit tag.

    Requires:
    - same normalized canonical key
    - tag present
    - no digest in raw ref
    """

    raw = (raw_ref or "").strip()
    canonical = (canonical_asset_key or "").strip().lower()
    if not raw or not canonical:
        return False
    normalized = normalize_container_ref(raw)
    if normalized.canonical_asset_key != canonical:
        return False
    return bool(normalized.observed_tag and not normalized.observed_digest)

File: frontend/lib/containerRefNormalization.ts
/**
 * Mirrors backend/app/services/container_ref_normalization.py and
 * infer_asset_kind(..., "") from asset_resolver.py for asset sidebar grouping.
 * Keep in sync when changing normalization rules.
 */

export interface NormalizedContainerRef {
  canonicalAssetKey: string;
  observedTag: string | null;
  observedDigest: string | null;
  rawRef: string;
}

function normalizeHexDigest(value: string | undefined): string | undefined {
  if (!value?.trim()) return undefined;
  const s = value.trim().toLowerCase();
  const hex = s.startsWith("sha256:")
    ? s.slice(7).replace(/[^0-9a-f]/g, "").slice(0, 64)
    : s.replace(/[^0-9a-f]/g, "").slice(0, 64);
  if (hex.length < 12) return undefined;
  return `sha256:${hex}`;
}

function splitDigest(ref: string): [string, string | null] {
  if (!ref.includes("@sha256:")) return [ref, null];
  const [left, right] = ref.split("@sha256:", 2);
  const digest = normalizeHexDigest(right ? `sha256:${right}` : undefined);
  return [left.trim(), digest ?? null];
}

function splitTag(repoRef: string): [string, string | null] {
  if (!repoRef.includes(":")) return [repoRef, null];
  const lastSlash = repoRef.lastIndexOf("/");
  const lastColon = repoRef.lastIndexOf(":");
  if (lastColon > lastSlash) {
    const tag = repoRef.slice(lastColon + 1).trim();
    if (tag) return [repoRef.slice(0, lastColon).trim(), tag];
  }
  return [repoRef, null];
}

function splitRegistryAndPath(repo: string): [string | null, string] {
  if (!repo) return [null, ""];
  if (!repo.includes("/")) return [null, repo];
  const first = repo.slice(0, repo.indexOf("/"));
  const rest = repo.slice(repo.indexOf("/") + 1);
  if (
    first.includes(".") ||
    first.includes(":") ||
    first === "localhost"
  ) {
    return [first, rest];
  }
  return [null, repo];
}

function dockerHubPath(path: string): string {
  const p = (path || "").replace(/^\/+|\/+$/g, "");
  if (!p) return "library/unknown";
  if (!p.includes("/")) return `library/${p}`;
  return p;
}

/**
 * Same semantics as backend normalize_container_ref().
 */
export function normalizeContainerRef(value: string | null | undefined): NormalizedContainerRef {
  const raw = (value ?? "").trim();
  if (!raw) {
    return {
      canonicalAssetKey: "docker.io/library/unknown",
      observedTag: null,
      observedDigest: null,
      rawRef: "",
    };
  }
  const [repoTagPart, digest] = splitDigest(raw);
  const [repoPart, tag] = splitTag(repoTagPart);
  let [registry, path] = splitRegistryAndPath(repoPart);

  if (registry === null) {
    registry = "docker.io";
    path = dockerHubPath(path);
  }

  let canonical = `${registry.toLowerCase()}/${(path || "").replace(/^\/+|\/+$/g, "").toLowerCase()}`.replace(/\/+$/g, "");
  if (!canonical) canonical = "docker.io/library/unknown";

  return {
    canonicalAssetKey: canonical,
    observedTag: tag || null,
    observedDigest: digest,
    rawRef: raw,
  };
}

/** Parse ``NEXT_PUBLIC_VAT_CONTAINER_ASSET_PATH_ALIASES``: ``from=>to;from2=>to2``. Empty ``to`` = strip ``from`` (same as backend). */
export function parseContainerAssetPathAliases(
  raw: string | undefined,
): Array<[string, string]> {
  const out: Array<[string, string]> = [];
  if (!raw?.trim()) return out;
  for (const segment of raw.split(";")) {
    const s = segment.trim();
    if (!s.includes("=>")) continue;
    const idx = s.indexOf("=>");
    const left = s.slice(0, idx);
    const right = s.slice(idx + 2);
    const src = left.trim().toLowerCase();
    const dst = right.trim().toLowerCase();
    if (src) out.push([src, dst]);
  }
  return out;
}

/**
 * Mirrors backend ``apply_container_asset_path_aliases`` (tenant prefix rewrite).
 * Read env on each call so tests can stub ``NEXT_PUBLIC_VAT_CONTAINER_ASSET_PATH_ALIASES``.
 */
export function applyContainerAssetPathAliases(canonicalKey: string): string {
  const k = canonicalKey.trim();
  if (!k) return canonicalKey;
  const raw =
    typeof process !== "undefined"
      ? (process.env.NEXT_PUBLIC_VAT_CONTAINER_ASSET_PATH_ALIASES ?? "")
      : "";
  const pairs = parseContainerAssetPathAliases(raw);
  if (!pairs.length) return canonicalKey;
  const lower = k.toLowerCase();
  for (const [src, dst] of pairs) {
    if (lower.startsWith(src)) {
      return dst + k.slice(src.length);
    }
  }
  return canonicalKey;
}

/**
 * Iron Bank VAT–style display: repository path **without** the registry host.
 * Canonical ref is `registry/org/.../image`; UI shows `org/.../image` (e.g.
 * `a1/caseprocessing/a1caseprocessing`, `containers/images/metrics-server`).
 * Identity remains the full normalized ref in `asset.id`; tags stay separate in UI.
 */
export function containerDisplayPathWithoutRegistry(ref: string): string {
  const { canonicalAssetKey } = normalizeContainerRef(ref);
  const parts = canonicalAssetKey.split("/").filter(Boolean);
  if (parts.length <= 1) {
    return parts[0] ?? canonicalAssetKey;
  }
  return parts.slice(1).join("/");
}

/**
 * Mirrors backend infer_asset_kind(asset_id, parser_id) with parser_id "".
 */
export function inferAssetKindForGrouping(
  assetId: string,
  parserId: string = "",
): string {
  const aid = (assetId || "").trim();
  const pid = (parserId || "").trim().toLowerCase();
  if (!aid) return "unknown";
  if (pid === "openscap" || pid === "openscap_oval") return "host_scope";
  if (pid === "semgrep" || pid === "sarif" || pid === "gitleaks")
    return "path_scope";
  if (pid === "npm_audit" || pid === "pip_audit") return "package_scope";
  if (aid.includes("/images/") || aid.startsWith("sha256:") || aid.includes(":"))
    return "container";
  if (aid.includes("/") && !aid.startsWith("/")) return "repo";
  if (aid.startsWith("/") || aid.startsWith("commit:") || aid.includes(">"))
    return "path_scope";
  return "package_scope";
}
